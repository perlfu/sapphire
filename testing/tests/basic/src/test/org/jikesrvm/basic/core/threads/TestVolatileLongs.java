/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.threads;

class TestVolatileLongs extends XThread {

  public static void main(String[] args) {
    VolatileLongField vlf = new VolatileLongField();
    for (int i = 0; i < 5; i++) {
      TestVolatileLongs tvl = new TestVolatileLongs(i,vlf);
      tvl.start();
    }
    XThread.say("bye");
    XThread.outputMessages();
  }

  static volatile int vi = 0;

  int n;
  long l;
  VolatileLongField vlf;

  TestVolatileLongs(int i, VolatileLongField vlf) {
    super("VL" + i);
    n = i;
    l = (((long) n) << 32) + n;
    this.vlf = vlf;
  }

  void performTask() {
    int errors = 0;
    for (int i = 0; i < 10000000; i++) {
      long tl = vlf.vl;
      vlf.vl = l;
      int n0 = (int) tl;
      int n1 = (int) (tl >> 32);
      if (n0 != n1) errors++;
      vi = n;
    }
    tsay(errors + " errors found");
  }

  private static class VolatileLongField {
    volatile long vl = 0;
  }

}
