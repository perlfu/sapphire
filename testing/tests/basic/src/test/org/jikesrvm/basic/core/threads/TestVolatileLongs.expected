Thread 0 Main: bye
Thread 1 VL0: creating
Thread 1 VL0: starting
Thread 1 VL0: run starting
Thread 1 VL0: 0 errors found
Thread 1 VL0: run finishing
Thread 2 VL1: creating
Thread 2 VL1: starting
Thread 2 VL1: run starting
Thread 2 VL1: 0 errors found
Thread 2 VL1: run finishing
Thread 3 VL2: creating
Thread 3 VL2: starting
Thread 3 VL2: run starting
Thread 3 VL2: 0 errors found
Thread 3 VL2: run finishing
Thread 4 VL3: creating
Thread 4 VL3: starting
Thread 4 VL3: run starting
Thread 4 VL3: 0 errors found
Thread 4 VL3: run finishing
Thread 5 VL4: creating
Thread 5 VL4: starting
Thread 5 VL4: run starting
Thread 5 VL4: 0 errors found
Thread 5 VL4: run finishing
